archive/tar.go

```diff
@@ -91,19 +92,62 @@ func AddFileToArchive(tw *tar.Writer, srcDir string, uid, gid int, fileSet map[s
 }
 
 func WriteTarFile(sourceDir, dest string, uid, gid int) (string, error) {
-	hasher := sha256.New()
 	f, err := os.Create(dest)
 	if err != nil {
 		return "", err
 	}
 	defer f.Close()
-	w := io.MultiWriter(hasher, f)
+
+	hasher := NewConcurrentHasher(sha256.New())
+	w := bufio.NewWriterSize(io.MultiWriter(hasher, f), 1024*1024)
 
 	if err := WriteTarArchive(w, sourceDir, uid, gid); err != nil {
 		return "", err
 	}
-	sha := hex.EncodeToString(hasher.Sum(make([]byte, 0, hasher.Size())))
-	return "sha256:" + sha, nil
+
+	if err := w.Flush(); err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf("sha256:%x", hasher.Sum(nil)), nil
+}
+
+type ConcurrentHasher struct {
```

Comment:
Would you mind moving the hasher implementation into a new file (e.g. `hash.go`)? I think that small organizational change would make both the hashing and the tarring logic more readable.